basement/Basement/UTF8/Types.hs

```diff
@@ -34,6 +36,9 @@ newtype StepDigit = StepDigit Word8
 -- | Step when processing ASCII character
 newtype StepASCII = StepASCII Word8
 
+-- | Specialized tuple used for case mapping.
+data CM = CM {-# UNPACK #-} !Char {-# UNPACK #-} !Char {-# UNPACK #-} !Char deriving (Eq)
```

Comment:
Should these be Char or Word8?

Reply:
Please document the invariant that it's isomorphic to `[Char]` and that all trailing `Char` must be `\0`.

Reply:
Suggest instead using `CM Int Int Int` with `-1` as the sentinel, since the rest of the foundation stuff is Int not Char for the UTFsize etc.